### PR TITLE
Refactor 'highlights'

### LINKED ISSUE #790: Decouple `highlights` and `panel` from the linting process
`highlights` and the `panel` view should listen for `events.FINISHED_LINTING` and act upon that

### LINKED ISSUE #902: error when setting same style for both error and warning
```
  File "/Users/koenlageveen/Library/Application Support/Sublime Text 3/Packages/SublimeLinter/sublime_linter.py", line 264, in highlight
    highlights.draw(view)
  File "/Users/koenlageveen/Library/Application Support/Sublime Text 3/Packages/SublimeLinter/lint/highlight.py", line 190, in draw
    "region keys not unique {}".format(drawn_regions)
AssertionError: region keys not unique ['sublime_linter.pylint.style_001', 'sublime_linter.pylint.gutter.style_001', 'sublime_linter.pylint.style_001', 'sublime_linter.pylint.gutter.style_001', 'sublime_linter.protected_regions']
```

When starting with these settings:

```json
{
    "gutter_theme": "default",
    "linters": {
        "pylint": {
            "styles": [
                {
                    "icon": "triangle",
                    "scope": "constant.numeric",
                    "types": ["error", "warning"]
                }
            ]
        }
    },
    "styles": [
        {
            "mark_style": "outline",
            "priority": 1,
            "scope": "region.yellowish markup.changed.sublime_linter markup.warning.sublime_linter",
            "icon": "triangle",
            "types": [
                "warning"
            ]
        },
        {
            "mark_style": "outline",
            "priority": 1,
            "scope": "region.redish markup.deleted.sublime_linter markup.error.sublime_linter",
            "icon": "error",
            "types": [
                "error"
            ]
        }
    ]
}
```

Splitting the pylint settings out into separate styles for error and warning seems to resolve the issue.

### PR DESCRIPTION
**Pull request by kaste**
Main goals: 

- Ensure that the *region_keys* are unique
- Optimize updating highlights by not just clearing everything before drawing
- Decouple data preparation code from side-effect code
- Prepare persistent data storage on view.settings()
- Decouple highlight.py from sublime_linter.py

Fixes #902 
Fixes #790 
Related #887

### CODE REVIEW

File: lint/highlight.py [hunk 1]
```diff
@@ -1,193 +0,0 @@
...
-
-class Highlight:
-    """This class maintains error marks and knows how to draw them."""
-
-    def __init__(self, view):
-        self.view = view
-
-        # Dict[error_type, Dict[style, List[region]]]
-        self.marks = defaultdict(lambda: defaultdict(list))
-        self.style_store = HighlightStyleStore()
```
Comment (pykong) [inline lint/highlight.py]: @kaste 
Just a quick note: Did you delete the class declaration as well?
As you deleted its only instantiation, it is not needed anymore.

Comment (kaste) [reply lint/highlight.py]: That's your job. 😁

Ahem. No.

File: highlight_view.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    by_position = defaultdict(list)
+    for error in errors:
+        line_start = get_line_start(view, error['line'])
+        region = sublime.Region(line_start + error['start'], line_start + error['end'])
+        by_position[(region.a, region.b)].append(ChainMap({'region': region}, error))
+
+    by_id = defaultdict(list)
+    for pos, errors in by_position.items():
+        # If we have multiple 'problems' here, 'error' takes precedence over
+        # 'warning'. We're lucky again that 'error' copmes before 'warning'
```
Comment (braver) [inline highlight_view.py]: Typo: copmes -> comes

Glad to be able to contribute something of value to this PR 💪🏻

**Comment by braver** [comment]
Works fine on my machine and indeed fixes #902. LGTM 🎉

File: lint/highlight.py (hunk 1 above)
Comment (pykong) [reply lint/highlight.py]: You kept it. I did not expect `HighlightStyleStore` to escape your wrath.

**Comment by kaste** [comment]
I just merge this with the typo fix. But I need to iterate on this towards #887

**Pull request merged**